src/norm_flows.jl

```diff
@@ -0,0 +1,139 @@
+using Distributions
+using LinearAlgebra
+using Random
+using StatsFuns: softplus
+using Roots, LinearAlgebra # for inverse
+
+################################################################################
+#                            Planar and Radial Flows                           #
+#             Ref: Variational Inference with Normalizing Flows,               #
+#               D. Rezende, S. Mohamed(2015) arXiv:1505.05770                  #
+################################################################################
+
+mutable struct PlanarLayer{T1,T2} <: Bijector
+    w::T1
+    u::T1
+    u_hat::T1
+    b::T2
+end
+
+function get_u_hat(u, w)
+    # To preserve invertibility
+    return (
+        u + (planar_flow_m(transpose(w) * u) - transpose(w) * u)[1]
+        * w / (norm(w[:,1],2) ^ 2)
+    ) # from A.1
+end
+
+function update_u_hat!(flow::PlanarLayer)
+    flow.u_hat = get_u_hat(flow.u, flow.w)
+end
+
+
+function PlanarLayer(dims::Int, container=Array)
+    w = container(randn(dims, 1))
+    u = container(randn(dims, 1))
+    b = container(randn(1))
+    u_hat = get_u_hat(u, w)
+    return PlanarLayer(w, u, u_hat, b)
+end
+
+planar_flow_m(x) = -1 .+ softplus.(x) # for planar flow from A.1
+dtanh(x) = 1 .- (tanh.(x)) .^ 2 # for planar flow
+ψ(z, w, b) = dtanh(transpose(w) * z .+ b) .* w # for planar flow from eq(11)
+
+function transform(flow::PlanarLayer, z)
+    return z + flow.u_hat * tanh.(transpose(flow.w) * z .+ flow.b) # from eq(10)
+end
+
+function forward(flow::T, z) where {T<:PlanarLayer}
+    update_u_hat!(flow)
+    # Compute log_det_jacobian
+    psi = ψ(z, flow.w, flow.b)
+    log_det_jacobian = log.(abs.(1.0 .+ transpose(psi) * flow.u_hat)) # from eq(12)
+    transformed = z + flow.u_hat * tanh.(transpose(flow.w) * z .+ flow.b)
+    return (rv=transformed, logabsdetjacob=log_det_jacobian) # from eq(10)
+end
+
+function inv(flow::PlanarLayer, y)
```

Comment:
1. add `u_hat = get_u_hat(flow.u, flow.w)`
2. change `flow.u_hat` to `u_hat` in this function.

We should not reply on the previous updated `flow.u_hat` as it is currently done. Why? It is possible that we never call `forward` but only inverse but still update the variable somehow (e.g. it is possible to back-prop through `inv` even there is an optimisation procedure inside.)

Reply:
should I remove u_hat as a field of PlanarLayer structure and instead calculate u_hat each time it is required?

Reply:
Yes exactly.